### DISCUSSION TITLE
Cannot Use CTEs in Custom Sql Query

### DISCUSSION THREAD

**Discussion by oluatte** [JasperFx/marten#2736] [Q&A]
Howdy and thanks for all the hard work on Marten.

I am resurrecting (somewhat dormant) Marten project and just upgraded to latest. All custom sql queries in the project that uses a CTE are broken with an exception like below...

 `Could not determine an 'id/Id' field or property for requested document type Api.Orders.Application.Handlers.GetOrganizationStatsQueryHandler+GetOrganizationStatsQueryResultSummaryPoco`
 
The problem might be this [line](https://github.com/JasperFx/marten/blob/21e172af3915123b1274926de98b516223b506bd/src/Marten/Linq/QueryHandlers/UserSuppliedQueryHandler.cs#L29C8-L29C97) which assumes every custom sql query starts with a `SELECT`. Queries using CTEs naturally start with `WITH` which fail the check and lead down the known / mapped T route. 

Is using CTEs in custom queries no longer allowed?

Thanks

**Comment by jeremydmiller** [comment]
I think that error means exactly what it says it does and has nothing to do with CTEs. At the moment, Marten is assuming that the top level document has some kind of identity, and it can't figure out what that would be for `GetOrganizationStatsQueryResultSummaryPoco`

**Reply by oluatte** [reply]
That makes sense to me. What I am pointing out is that the logic as-is excludes the use of CTEs (because the SQL won't start with "SELECT), and that may not have been intentional (since it used to work in previous versions). 

Marten's Linq support is great (and improving based on the current roadmap), but it is critical to have a good escape hatch for complex queries. In this case, any number of fixes would work:

- Checking more starting keywords than `SELECT`
- Overloads of the Query methods that allow specificying that the query is custom and bypassing any starting keyword logic.
-  Parallel methods that bypass starting keyword logic 
-  etc.

**Reply by kseitz10** [reply]
> Overloads of the Query methods that allow specifying that the query is custom and bypassing any starting keyword logic.

I would personally support this, since I recently fell into the same trap. (I brought this up on Discord having checked the issues section of this repo, but not discussions.) This was a surprise for us during the v6 upgrade. We could PR a change to make these APIs more explicit in how they behave.
